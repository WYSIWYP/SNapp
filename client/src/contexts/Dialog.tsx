import React, {createContext, useContext, useReducer, useRef} from 'react';

export type state = {
    shown: boolean,
    data?: {
        title: string,
        contents: any,
        width: number,
        height: number,
    }
};
export type action = {
    type: 'open', val: state['data'],
} | {
    type: 'close',
};

let initialState: state = {
    shown: false,
};

export const DialogContext = createContext(undefined! as [state, React.Dispatch<action>]);
export const DialogStateProvider: React.FC<{}> = ({children}) => {
    //we need to use a ref here to ensure that the same reducer is always used
    let reducer = useRef((state: state, action: action): state => {
        switch (action.type) {
            case 'open':
                return {
                    shown: true,
                    data: action.val,
                };
            case 'close':
                return {shown: false};
        }
    });
    let [state, dispatch] = useReducer(reducer.current, initialState);
    return (
        <DialogContext.Provider value={[state, dispatch]}>
            {children}
        </DialogContext.Provider>
    );
};
export const useDialogState = () => useContext(DialogContext);
