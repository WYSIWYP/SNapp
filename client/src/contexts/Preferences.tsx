import React, {createContext, useContext, useReducer, useRef} from "react";

export const colorPreferenceStyles = {
    black: "#000000",
    grey: "#777777",
    red: "#FF0000",
    orange: "orange",
    yellow: "#FFFF00",
    green: "#00FF00",
    blue: "#0000FF",
    purple: "purple"
};
export const colorPreferenceOptions = Object.keys(
    colorPreferenceStyles
) as (keyof typeof colorPreferenceStyles)[];
export type colorPreferenceOption = (typeof colorPreferenceOptions)[number];

export const scalePreferenceOptions = ['small', 'medium', 'large'] as const;
export type scalePreferenceOption = (typeof scalePreferenceOptions)[number];

export const spacingPreferenceOptions = ['narrow', 'moderate', 'wide'] as const;
export type spacingPreferenceOption = (typeof spacingPreferenceOptions)[number];

export const naturalNoteHeadPreferenceOptions = ["●", "○"] as const; 
export type naturalNoteHeadPreferenceOption = (typeof naturalNoteHeadPreferenceOptions)[number];

export const sharpNoteHeadPreferenceOptions = ["▲", "△", "#"] as const;
export type sharpNoteHeadPreferenceOption = (typeof sharpNoteHeadPreferenceOptions)[number];

export const flatNoteHeadPreferenceOptions = ["▼", "▽", "b"] as const; 
export type flatNoteHeadPreferenceOption = (typeof flatNoteHeadPreferenceOptions)[number];

export const clefPreferenceOptions = ["WYSIWYP","Traditional"] as const;
export type clefPreferenceOptions = (typeof clefPreferenceOptions)[number];

export const measuresPerRowOptions = [1, 2, 3, 4, 5, 6, 7, 8] as const; // TODO: Consider using a slider  // 21 June 2021 increase to 8
export type measuresPerRowOption = (typeof measuresPerRowOptions)[number];

export const accidentalTypeOptions = ['auto', 'sharps', 'flats'] as const;
export type accidentalTypeOption = (typeof accidentalTypeOptions)[number];

export const lyricsFontSizeOptions = ['small', 'medium', 'large'] as const;
export type lyricsFontSizeOption = (typeof lyricsFontSizeOptions)[number];

export type state = {
    noteDurationColor: colorPreferenceOption;
    noteSymbolColor: colorPreferenceOption;
    staffScale: scalePreferenceOption;
    horizontalSpacing: spacingPreferenceOption;
    verticalSpacing: spacingPreferenceOption;
    noteScale: scalePreferenceOption;
    naturalNoteShape: naturalNoteHeadPreferenceOption,
    sharpNoteShape: sharpNoteHeadPreferenceOption;
    flatNoteShape: flatNoteHeadPreferenceOption;
    clefSymbols: clefPreferenceOptions;
    measuresPerRow: measuresPerRowOption;
    accidentalType: accidentalTypeOption;
    lyricsFontSize: lyricsFontSizeOption
};
export type action = {
    type: "set";
    val: Partial<state>;
};

let initialState: state = {
    noteDurationColor: "grey",
    noteSymbolColor: "black",
    staffScale: 'medium',
    horizontalSpacing: 'narrow',  // 21 June 2021  changed from medium
    verticalSpacing: 'narrow',    // 21 June 2021  changed from medium
    noteScale: 'medium',
    naturalNoteShape: '●',
    sharpNoteShape: '▲',
    flatNoteShape: '▼',
    clefSymbols: 'WYSIWYP',
    measuresPerRow: 4,
    accidentalType: 'auto',
    lyricsFontSize: 'small'
};

export const PreferencesContext = createContext(undefined! as [
    state,
    React.Dispatch<action>
]);
export const PreferencesStateProvider: React.FC<{}> = ({children}) => {
    //we need to use a ref here to ensure that the same reducer is always used
    let reducer = useRef(
        (state: state, action: action): state => {
            switch (action.type) {
                case "set":
                    return {...state, ...action.val};
            }
        }
    );
    let [state, dispatch] = useReducer(reducer.current, initialState);
    return (
        <PreferencesContext.Provider value={[state, dispatch]}>
            {children}
        </PreferencesContext.Provider>
    );
};
export const usePreferencesState = () => useContext(PreferencesContext);
