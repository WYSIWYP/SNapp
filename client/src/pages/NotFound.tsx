import React, {CSSProperties} from 'react';
import {RouteComponentProps} from "@reach/router";

type Props = {} & RouteComponentProps;

const NotFound: React.FC<Props> = () => {
    return (
        <div style={styles.body}>
            That page could not be found
        </div>
    );
};

const styleMap = {
    body: {
        top: '50%',
        height: 'auto',
        transform: 'translate(0px,-50px)',
        textAlign: 'center',
        fontSize: '22px',
    },
} as const;
const styles: Record<keyof typeof styleMap,CSSProperties> = styleMap;

export default NotFound;
