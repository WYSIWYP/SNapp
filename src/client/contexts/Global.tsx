import React from 'react';

const GlobalStateProvider: React.FC<{}> = ({children}) => {
    //we need to use a ref here to ensure that the same reducer is always used
    let stateProviders = [] as React.FunctionComponent[];
    stateProviders.forEach(x=>{
        children = x({children});
    });
    return <>{children}</>;
};
export default GlobalStateProvider;
