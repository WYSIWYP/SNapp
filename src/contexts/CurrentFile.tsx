import React, {createContext, useContext, useReducer, useRef} from 'react';
import MusicXML from 'musicxml-interfaces';

export type state = {
    id?: string,
    file_name?: string,
    data?: MusicXML.ScoreTimewise
};
export type action = {
    type: 'set', val: state,
};

let initialState: state = {};

type Props = {
    children?: any,
};

export const CurrentFileContext = createContext(undefined! as [state, React.Dispatch<action>]);
export const CurrentFileStateProvider: React.FC<Props> = ({children}) => {
    //we need to use a ref here to ensure that the same reducer is always used
    let reducer = useRef((state: state, action: action): state => {
        switch (action.type) {
            case 'set':
                return action.val;
        }
    });
    let [state, dispatch] = useReducer(reducer.current, initialState);
    return (
        <CurrentFileContext.Provider value={[state, dispatch]}>
            {children}
        </CurrentFileContext.Provider>
    );
};
export const useCurrentFileState = () => useContext(CurrentFileContext);
