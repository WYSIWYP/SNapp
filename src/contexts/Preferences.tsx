import React, {createContext, useContext, useReducer, useRef} from "react";

export const colorPreferenceStyles = {
    black: "#000000",
    grey: "#777777",
    red: "#FF0000",
    orange: "orange",
    yellow: "#FFFF00",
    green: "#00FF00",
    blue: "#0000FF",
    purple: "purple"
};
export const colorPreferenceOptions = Object.keys(
    colorPreferenceStyles
) as (keyof typeof colorPreferenceStyles)[];
export type colorPreferenceOption = (typeof colorPreferenceOptions)[number];

export const scalePreferenceOptions = ['small', 'medium', 'large'] as const;
export type scalePreferenceOption = (typeof scalePreferenceOptions)[number];

export const spacingPreferenceOptions = ['narrow', 'moderate', 'wide'] as const;
export type spacingPreferenceOption = (typeof spacingPreferenceOptions)[number];

export const naturalNoteHeadPreferenceOptions = ["● filled", "○ hollow"] as const; 
export type naturalNoteHeadPreferenceOption = (typeof naturalNoteHeadPreferenceOptions)[number];

export const sharpNoteHeadPreferenceOptions = ["▲ filled", "△ hollow", "▀ combo", "#"] as const;                 // 24 July 2022 added half box sharp
export type sharpNoteHeadPreferenceOption = (typeof sharpNoteHeadPreferenceOptions)[number];  

export const flatNoteHeadPreferenceOptions = ["▼ filled", "▽ hollow", "▄ combo", "b"] as const;                  // 24 July 2022 added half box flat 
export type flatNoteHeadPreferenceOption = (typeof flatNoteHeadPreferenceOptions)[number];

export const clefPreferenceOptions = ["WYSIWYP","Traditional"] as const;
export type clefPreferenceOption = (typeof clefPreferenceOptions)[number];

export const measuresPerRowOptions = [1, 2, 3, 4, 5, 6, 7, 8] as const; // TODO: Consider using a slider  // 21 June 2021 increase to 8
export type measuresPerRowOption = (typeof measuresPerRowOptions)[number];

export const accidentalTypeOptions = ['auto', 'sharps', 'flats'] as const;
export type accidentalTypeOption = (typeof accidentalTypeOptions)[number];

export const lyricsFontSizeOptions = ['small', 'medium', 'large'] as const;
export type lyricsFontSizeOption = (typeof lyricsFontSizeOptions)[number];

export const fingeringsPreferenceOptions = ['hide', 'above', 'behind', 'left'] as const;
export type fingeringsOption = (typeof fingeringsPreferenceOptions)[number];

export const cccPreferenceOptions = ['enable', 'disable'] as const;
export type cccOption = (typeof cccPreferenceOptions)[number];

export type state = {
    noteDurationColor: colorPreferenceOption;
    noteSymbolColor: colorPreferenceOption;
    staffScale: scalePreferenceOption;
    horizontalSpacing: spacingPreferenceOption;
    verticalSpacing: spacingPreferenceOption;
    noteScale: scalePreferenceOption;
    naturalNoteShape: naturalNoteHeadPreferenceOption,
    sharpNoteShape: sharpNoteHeadPreferenceOption;
    flatNoteShape: flatNoteHeadPreferenceOption;
    clefSymbols: clefPreferenceOption;
    measuresPerRow: measuresPerRowOption;
    accidentalType: accidentalTypeOption;
    lyricsFontSize: lyricsFontSizeOption;
    fingeringsDisplay: fingeringsOption;
    cccDisplay: cccOption
};
export type action = {
    type: "set";
    val: Partial<state>;
};

let initialState: state = {
    noteDurationColor: "grey",
    noteSymbolColor: "black",
    staffScale: 'medium',
    horizontalSpacing: 'narrow',  
    verticalSpacing: 'narrow',    
    noteScale: 'medium',
    naturalNoteShape: '○ hollow',
    sharpNoteShape: '▀ combo',
    flatNoteShape: '▄ combo',
    clefSymbols: 'WYSIWYP',
    measuresPerRow: 5,
    accidentalType: 'auto',
    lyricsFontSize: 'small',
    fingeringsDisplay: 'hide', 
    cccDisplay: 'disable'
};

export const PreferencesContext = createContext(undefined! as [
    state,
    React.Dispatch<action>
]);

type Props = {
    children?: any,
};

export const PreferencesStateProvider: React.FC<Props> = ({ children }) => {
    //we need to use a ref here to ensure that the same reducer is always used
    let reducer = useRef(
        (state: state, action: action): state => {
            switch (action.type) {
                case "set":
                    return {...state, ...action.val};
            }
        }
    );
    let [state, dispatch] = useReducer(reducer.current, initialState);
    return (
        <PreferencesContext.Provider value={[state, dispatch]}>
            {children}
        </PreferencesContext.Provider>
    );
};
export const usePreferencesState = () => useContext(PreferencesContext);
