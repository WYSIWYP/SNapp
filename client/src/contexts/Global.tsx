import React from 'react';
import {DialogStateProvider} from './Dialog';
import {CurrentFileStateProvider} from './CurrentFile';
import {PreferencesStateProvider} from './Preferences';

const GlobalStateProvider: React.FC<{}> = ({children}) => {
    //we need to use a ref here to ensure that the same reducer is always used
    let stateProviders = [DialogStateProvider, CurrentFileStateProvider, PreferencesStateProvider] as React.FunctionComponent[];
    stateProviders.forEach(x=>{
        children = x({children});
    });
    return <>{children}</>;
};
export default GlobalStateProvider;
