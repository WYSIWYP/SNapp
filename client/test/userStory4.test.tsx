import React from 'react';
import {shallow} from 'enzyme';

import {PreferencesStateProvider} from '../src/contexts/Preferences';
import NotFound from '../src/pages/NotFound';

it('renders PreferencesStateProvider and Router correctly', () => {
    let tree = shallow(<PreferencesStateProvider />);
    expect(tree).toMatchSnapshot();
});



it('renders not found page correctly ', () => {
    let tree = shallow(<NotFound />);
    expect(tree).toMatchSnapshot();
});

