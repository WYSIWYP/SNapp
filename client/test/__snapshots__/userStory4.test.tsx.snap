// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders PreferencesStateProvider and Router correctly 1`] = `
<ContextProvider
  value={
    Array [
      Object {
        "flatNoteShape": "▼",
        "horizontalSpacing": 2,
        "measuresPerRow": 4,
        "naturalNoteShape": "●",
        "noteDurationColor": "grey",
        "noteScale": "medium",
        "noteSymbolColor": "black",
        "sharpNoteShape": "▲",
        "staffScale": "medium",
        "verticalSpacing": 2,
      },
      [Function],
    ]
  }
/>
`;

exports[`renders not found page correctly  1`] = `
<div
  style={
    Object {
      "fontSize": "22px",
      "height": "auto",
      "textAlign": "center",
      "top": "50%",
      "transform": "translate(0px,-50px)",
    }
  }
>
  That page could not be found
</div>
`;
